homeassistant/components/hive/config_flow.py

```diff
@@ -0,0 +1,184 @@
+
+            # Login to the Hive.
+            try:
+                self.tokens = await self.hive_auth.login()
+            except HiveInvalidUsername:
+                errors["base"] = "invalid_username"
+            except HiveInvalidPassword:
+                errors["base"] = "invalid_password"
+            except HiveApiError:
+                errors["base"] = "no_internet_available"
+
+            if self.tokens.get("ChallengeName") == "SMS_MFA":
+                # Complete SMS 2FA.
+                return await self.async_step_2fa()
+
+            if "AuthenticationResult" in self.tokens:
+                # Complete the entry setup.
+                try:
+                    return await self.async_setup_hive_entry()
+                except UnknownHiveError:
+                    errors["base"] = "unknown"
+
+        # Show User Input form.
+        schema = vol.Schema(
+            {vol.Required(CONF_USERNAME): str, vol.Required(CONF_PASSWORD): str}
+        )
+        return self.async_show_form(step_id="user", data_schema=schema, errors=errors)
+
+    async def async_step_2fa(self, user_input=None):
+        """Handle 2fa step."""
+        errors = {}
+        result = {}
+
+        if user_input and user_input["2fa"] == "0000":
+            self.tokens = await self.hive_auth.login()
+        elif user_input:
+
+            try:
+                result = await self.hive_auth.sms_2fa(user_input["2fa"], self.tokens)
+            except HiveInvalid2FACode:
+                errors["base"] = "invalid_code"
+            except HiveApiError:
+                errors["base"] = "no_internet_available"
+
+            if "AuthenticationResult" in result:
+                self.tokens = result
+                try:
+                    return await self.async_setup_hive_entry()
+                except UnknownHiveError:
+                    errors["base"] = "unknown"
+
+        schema = vol.Schema({vol.Required(CONF_CODE): str})
+        return self.async_show_form(step_id="2fa", data_schema=schema, errors=errors)
+
+    async def async_setup_hive_entry(self):
+        """Finish setup and create the config entry."""
+
+        if "AuthenticationResult" not in self.tokens:
+            raise UnknownHiveError
+
+        # Setup the config entry
+        self.data["tokens"] = self.tokens
+        if self.entry:
+            self.hass.config_entries.async_update_entry(
+                self.entry, title=self.data["username"], data=self.data
+            )
+            await self.hass.config_entries.async_reload(self.entry.entry_id)
+            return self.async_abort(reason="reauth_successful")
+        return self.async_create_entry(title=self.data["username"], data=self.data)
+
+    async def async_step_reauth(self, user_input=None):
+        """Re Authenticate a user."""
+        errors = {}
+
+        if user_input is not None:
+            for entry in self._async_current_entries():
+                if entry.unique_id == self.unique_id:
```

Comment:
We can set the unique id of the flow already when starting it by passing that item in the `context` parameter. Then we can simplify here and just pass the user input to the user step.

Reply:
Im sorry I don't get what you mean with this one?